### PR TITLE
transform shaders into correct coordinate space

### PR DESCRIPTION
**Pull request by jonahwilliams**
TBD tests.

image

### CODE REVIEW

File: packages/vector_graphics_compiler/lib/src/geometry/path.dart [hunk 1]
```diff
@@ -453,6 +455,41 @@ class Path {
...
+          break;
+        case PathCommandType.line:
+          final LineToCommand move = command as LineToCommand;
+          smallestX = math.min(move.x, smallestX);
+          smallestY = math.min(move.y, smallestY);
+          largestX = math.max(move.x, largestX);
+          largestY = math.max(move.y, largestY);
+          break;
+        case PathCommandType.cubic:
+          // TODO: Handle this case.
```
Comment (jonahwilliams) [inline packages/vector_graphics_compiler/lib/src/geometry/path.dart:484]: Need to look up how skia does this

Comment (jonahwilliams) [reply packages/vector_graphics_compiler/lib/src/geometry/path.dart:484]: https://bugs.chromium.org/p/skia/issues/detail?id=5751

Seems like we should just use the bounds that contain all control points

Comment (dnfield) [reply packages/vector_graphics_compiler/lib/src/geometry/path.dart:484]: Impeller has easier code for this too.

I'm slightly of the mind of exposing a bunch of Skia's path stuff (i.e. bounds calculation, tight bounds calculation, path operations) in a C++ lib and consuming that from this, similar to the plan for tessellation.

File: packages/vector_graphics_compiler/lib/src/paint.dart [hunk 2]
```diff
@@ -136,6 +139,23 @@ class LinearGradient extends Shader {
...
 
+  @override
+  LinearGradient applyBounds(Rect bounds, AffineMatrix transform) {
+    return LinearGradient(
+      from: transform.transformPoint(
+          Point(from.x * bounds.width, from.y * bounds.height) +
+              Point(bounds.left, bounds.top)),
+      to: transform.transformPoint(
+          Point(to.x * bounds.width, to.y * bounds.height) +
+              Point(bounds.left, bounds.top)),
```
Comment (dnfield) [inline packages/vector_graphics_compiler/lib/src/paint.dart:150]: nit: more commas

File: packages/vector_graphics_compiler/lib/src/paint.dart [hunk 3]
```diff
@@ -243,6 +263,26 @@ class RadialGradient extends Shader {
...
+      radius: radius,
+      colors: colors,
+      offsets: offsets,
+      tileMode: tileMode,
+      transform: this.transform,
+      focalPoint: focalPoint == null
+          ? focalPoint
+          : transform.transformPoint(Point(
+                  focalPoint!.x * bounds.width, focalPoint!.y * bounds.height) +
+              Point(bounds.left, bounds.top)),
```
Comment (dnfield) [inline packages/vector_graphics_compiler/lib/src/paint.dart:281]: same nit

File: packages/vector_graphics_compiler/lib/src/svg/node.dart [hunk 4]
```diff
@@ -147,6 +147,7 @@ class ParentNode extends Node {
   void build(DrawCommandBuilder builder, AffineMatrix transform) {
     final bool requiresLayer = _requiresSaveLayer();
     if (requiresLayer) {
+      // TODO: bounds
```
Comment (dnfield) [inline packages/vector_graphics_compiler/lib/src/svg/node.dart:150]: I think we can remove this. If you dn't specify a layer, Skia currently just uses the current clip, which is as good as or better than whatever we'd calculate here.

**Review by dnfield — APPROVED** [review]
Very nice, LGTM with nits

**Comment by dnfield** [comment]
Oh.

I guess we should add tests. Like, probably tests for bounds and tests for point.+ and tests for shaders getting transformed to the right coordinate space...

File: packages/vector_graphics_compiler/lib/src/geometry/path.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/vector_graphics_compiler/lib/src/geometry/path.dart:484]: depends on how complicated it is to add cubics. Currently this code is highly trivial

File: packages/vector_graphics_compiler/lib/src/svg/node.dart (hunk 4 above)
Comment (jonahwilliams) [reply packages/vector_graphics_compiler/lib/src/svg/node.dart]: done

File: packages/vector_graphics_compiler/lib/src/paint.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/vector_graphics_compiler/lib/src/paint.dart]: done

**Review by dnfield — APPROVED** [review]
This still LGTM, but bonus points if you can do something that tests that we actually call gradient.applyBounds when we create the command list

**Comment by jonahwilliams** [comment]
eventually, sure :)

**Pull request merged**